src/ripple/app/paths/impl/PaymentView.h

```diff
@@ -18,28 +18,47 @@
 //==============================================================================
 
 #ifndef RIPPLE_APP_PAYMENTVIEW_H_INCLUDED
+#define RIPPLE_APP_PAYMENTVIEW_H_INCLUDED
 
+#include <ripple/core/Config.h>
+#include <ripple/app/ledger/MetaView.h>
 #include <ripple/ledger/View.h>
+#include <ripple/ledger/ViewAPIBasics.h>
 #include <ripple/ledger/DeferredCredits.h>
+#include <utility>
 
 namespace ripple {
 
 /** A View wrapper which makes credits unavailable to balances.
 
-    This is used for payments, so that consuming liquidity
-    from a path never causes portions of that path or other
-    paths to gain liquidity.
+    This is used for payments and pathfinding, so that consuming
+    liquidity from a path never causes portions of that path or
+    other paths to gain liquidity.
+
+    For this to work, callers need to go through member functions
+    like accountFunds instead of the free functions, or else the
+    balances will be incorrect.
 */
 class PaymentView : public View
 {
 private:
-    View& view_;
+    MetaView view_;
     DeferredCredits tab_;
+    PaymentView* pv_ = nullptr;
 
 public:
+    template <class... Args>
+    explicit
+    PaymentView (Args&&... args)
+        : view_ (std::forward<Args>(args)...)
+    {
+    }
+
     explicit
-    PaymentView (View& view)
-        : view_(view)
+    PaymentView (PaymentView& parent)
+        : view_ (parent,
+            parent.openLedger())
+        , pv_ (&parent)
     {
```

Comment:
`PaymentView` "attaches" to its parent, which must be a `PaymentView` if you go through that particular constructor. The views form a linked list with the most recent one at the front, and subsequent views accessible through a call to `parent()`. The constructor does not destroy its parent parameter. The function `PaymentView::apply` applies changes to the parent. After a call to `apply`, the only valid operation on a `PaymentView` is to call the destructor.